cmd/nex/rootfs.go

```diff
@@ -1,3 +1,389 @@
 package main
 
-type RootFS struct{}
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+type reference struct {
+	RawRef     string
+	Repo       *remote.Repository
+	Descriptor v1.Descriptor
+	Manifest   *v1.Manifest
+}
+
+type RootFS struct {
+	OciRef    string            `arg:"" name:"oci-ref" help:"OCI image reference"`
+	Files     map[string]string `placeholder:"src=dest;..." help:"Add a file to the rootfs"`
+	Os        string            `help:"Set the OS to use when parsing the OCI reference" default:"linux"`
+	Arch      string            `help:"Set the architecture to use when parsing the OCI reference" default:"amd64"`
+	Status    bool              `help:"Print the status"`
+	OutputDir string            `help:"Directory to output rootfs into" default:"."`
+	NoTLS     bool              `help:"Disable TLS when connecting to registry" default:"false"`
+
+	ociRef   *reference `kong:""`
+	buildDir string     `kong:""`
+}
+
+func (r *RootFS) Run() error {
+	var err error
+
+	if os.Getuid() != 0 {
+		return errors.New("Please run as root")
+	}
+
+	r.buildDir, err = os.MkdirTemp(os.TempDir(), "rootfs-*")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(r.buildDir)
+
+	r.ociRef = &reference{
+		RawRef: r.OciRef,
+	}
+
+	err = r.Initialize()
+	if err != nil {
+		return err
+	}
+
+	err = r.downloadExtractLayers()
+	if err != nil {
+		return err
+	}
+
+	for src, dst := range r.Files {
+		err = r.AddFile(src, dst)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = r.makeRootFS()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *RootFS) Initialize() error {
+	repo, err := remote.NewRepository(r.ociRef.RawRef)
+	if err != nil {
+		return err
+	}
+	if r.NoTLS {
+		repo.PlainHTTP = true
+	}
+	r.ociRef.Repo = repo
+
+	repo.Client = &auth.Client{
+		Cache: auth.DefaultCache,
+		Credential: func(ctx context.Context, registry string) (auth.Credential, error) {
+			switch r.ociRef.Repo.Reference.Registry {
+			case "docker.io":
+				authGet, err := http.Get(fmt.Sprintf("https://auth.docker.io/token?service=registry.docker.io&scope=repository:%s:pull", repo.Reference.Repository))
```

Comment:
Helper function to return the URL might be nice.

Reply:
I decided against it because it's only for docker.io and the URL is always the same.  I can pull it up though to make it easier to read